### PR TITLE
Replace orddict by maps in sys_core_fold

### PR DESCRIPTION
**Pull request by josevalim**
The following command was run before and after the change:

    erlc +'{eprof,sys_core_fold}' -S src/core_parse.erl

Before the patch, the last entries were:

    orddict:store/3                        22065     1.66    9219  [0.42]
    cerl:is_data/1                         22229     1.66    9221  [0.41]
    sys_core_fold:'-pattern_list/3-anon    25212     1.78    9886  [0.39]
    sys_core_fold:'-case_will_var_match    23691     1.81   10031  [0.42]
    cerl:get_ann/1                         24717     1.82   10094  [0.41]
    cerl_clauses:match/2                   26508     1.89   10518  [0.40]
    sys_core_fold:pattern/3                47973     2.30   12745  [0.27]
    erlang:setelement/3                    62120     2.44   13554  [0.22]
    sys_core_fold:'-case_opt_args/5-lc$    33781     2.53   14069  [0.42]
    cerl_clauses:match_list/3              53296     2.63   14619  [0.27]
    sys_core_fold:sub_del_var/2            39264     2.88   15973  [0.41]
    sys_core_fold:sub_is_val/2             39599     2.91   16167  [0.41]
    lists:mapfoldl/3                       47003     3.13   17350  [0.37]
    cerl_clauses:match/3                   66962     3.78   20998  [0.31]
    sys_core_fold:expr/3                   67430     4.33   24065  [0.36]
    orddict:find/2                         93914     5.00   27731  [0.30]
    cerl_sets:add_element/2                44338     5.30   29422  [0.66]
    cerl:type/1                            70342     5.62   31191  [0.44]
    cerl_sets:is_element/2                 78865     5.78   32102  [0.41]
    -----------------------------------  -------  -------  ------  [----]
    Total:                               1512026  100.00%  555175  [0.37]

After the patch, they were:

    cerl:is_data/1                         22229     1.71    8953  [0.40]
    sys_core_fold:'-case_will_var_match    23691     1.76    9219  [0.39]
    sys_core_fold:'-pattern_list/3-fun-    25402     1.81    9500  [0.37]
    cerl_clauses:match/2                   26508     1.83    9575  [0.36]
    maps:put/3                              9711     1.86    9749  [1.00]
    cerl:get_ann/1                         24717     2.01   10535  [0.43]
    sys_core_fold:pattern/3                47973     2.37   12432  [0.26]
    erlang:setelement/3                    61668     2.57   13475  [0.22]
    sys_core_fold:'-case_opt_args/5-lc$    33781     2.58   13500  [0.40]
    cerl_clauses:match_list/3              53296     2.74   14352  [0.27]
    sys_core_fold:sub_del_var/2            39264     2.90   15168  [0.39]
    sys_core_fold:sub_is_val/2             39599     2.97   15574  [0.39]
    lists:mapfoldl/3                       47003     3.17   16616  [0.35]
    cerl_clauses:match/3                   66962     3.86   20197  [0.30]
    sys_core_fold:'-expr_list/3-lc$^0/1    46165     4.04   21156  [0.46]
    cerl_sets:add_element/2                44338     4.47   23408  [0.53]
    sys_core_fold:expr/3                   67430     4.49   23513  [0.35]
    cerl:type/1                            70342     5.49   28720  [0.41]
    cerl_sets:is_element/2                 78865     5.97   31266  [0.40]
    -----------------------------------  -------  -------  ------  [----]
    Total:                               1464895  100.00%  523594  [0.36]

+time was also used but we could not measure a relevant difference.

### CODE REVIEW

File: lib/compiler/src/sys_core_fold.erl [hunk 1]
```diff
@@ -1405,7 +1406,7 @@ is_subst(_) -> false.
 %%
 %%  We use the variable name as key so as not have problems with
 %%  annotations.  When adding a new substitute we fold substitute
-%%  chains so we never have to search more than once.  Use orddict so
+%%  chains so we never have to search more than once.  Use map so
```
Comment (bjorng) [inline lib/compiler/src/sys_core_fold.erl]: `"Use map so we know the format."` does not make much sense. I suggest you delete the sentence.

File: lib/compiler/src/sys_core_fold.erl [hunk 2]
```diff
@@ -1444,40 +1445,34 @@ sub_del_var(#c_var{name=V}, #sub{v=S,s=Scope,t=Tdb}=Sub) ->
...
-              end,
-    S = sub_subst_scope_1(cerl_sets:to_list(Scope), Initial, S0),
-    Sub#sub{v=orddict:from_list(S)}.
-
-%% The keys in an orddict must be unique. Make them so!
+sub_subst_scope(#sub{v=V0,s=Scope,c=C0}=Sub) ->
+    {C,V} = sub_subst_scope_1(cerl_sets:to_list(Scope), C0, V0),
+    Sub#sub{v=V,c=C}.
+
+%% The keys in an maps must be unique. Make them so!
```
Comment (bjorng) [inline lib/compiler/src/sys_core_fold.erl]: "an maps" => "a map"

File: lib/compiler/src/sys_core_fold.erl [hunk 3]
```diff
@@ -1444,40 +1445,34 @@ sub_del_var(#c_var{name=V}, #sub{v=S,s=Scope,t=Tdb}=Sub) ->
...
+    sub_subst_scope_1(T, Key-1, maps:put(Key, #c_var{name=H}, Acc));
+sub_subst_scope_1([], Key, Acc) -> {Key,Acc}.
 
 sub_is_val(#c_var{name=V}, #sub{v=S,s=Scope}) ->
     %% When the bottleneck in sub_del_var/2 was eliminated, this
     %% became the new bottleneck. Since the scope contains all
     %% live variables, a variable V can only be the target for
     %% a substitution if it is in the scope.
-    cerl_sets:is_element(V, Scope) andalso v_is_value(V, S).
+    cerl_sets:is_element(V, Scope) andalso v_is_value(V, maps:to_list(S)).
```
Comment (bjorng) [inline lib/compiler/src/sys_core_fold.erl]: I am a little bit worried about this change. Most of the time, `V` is not in the scope, so there is no need to do `maps:to_list/1`. But when `V` is in the scope, this operations becomes much more expensive.

Comment (josevalim) [inline lib/compiler/src/sys_core_fold.erl]: @bjorng Yes, this is the part that worries me too. I did benchmark two or three big modules in Elixir and I also could not measure anything relevant. Profiling showed improvements but they did not translate to actual improvements - as above.

The best would be if we could rewrite part of this code to rely on map lookups but at the moment I don't have enough knowledge of this code to devise alternatives.

Maybe the best course of action for now is to close this pull request. I can revisit it later once I am more familiar with sys_core_fold.

**Comment by bjorng** [comment]
> Maybe the best course of action for now is to close this pull request. I can revisit it later once I am more familiar with sys_core_fold.

Yes, I think that is a good idea.

It is also possible that we will optimize some of the functions in `maps` (e.g. `maps:fold/3` so that it does not have convert the map to a list) during the course of OTP 21 development.

**Pull request closed without merging**